File: JS/webgl_data/glcommon.js
import { isNullOrWhiteSpace, interpolate, toText, split } from "../.fable/fable-library.3.0.0/String.js";
import { append as append_1, tryFind } from "../.fable/fable-library.3.0.0/Array.js";
import { reverse, tryFind as tryFind_2, singleton, delay, append, collect } from "../.fable/fable-library.3.0.0/Seq.js";
import { cons, filter, singleton as singleton_1, append as append_2, empty, sortBy, tryFind as tryFind_1 } from "../.fable/fable-library.3.0.0/List.js";
import { Data_GlAttributeData__get_Name, Data_GlUniformData__get_Name, Data_GlUboData__get_Name } from "./webgl_data.js";
import { comparePrimitives } from "../.fable/fable-library.3.0.0/Util.js";

export function splitName(name) {
    const pair = split(name, ["."], null, 0);
    return [pair[0], pair[1]];
}

export function getScene(name, data) {
    const matchValue = tryFind((u) => (u.Name === name), data.Scenes);
    if (matchValue == null) {
        throw (new Error(toText(interpolate("Scene \u0027%P()\u0027 not found", [name]))));
    }
    else {
        const scene = matchValue;
        return scene;
    }
}

export function allObjects(data) {
    const sceneObjects = (data_1) => collect((l) => l.Objects, data_1.Layers);
    return collect(sceneObjects, data.Scenes);
}

export function getAttributeData(attribute) {
    let pattern_matching_result, a;
    if (attribute.tag === 1) {
        pattern_matching_result = 0;
        a = attribute.fields[0];
    }
    else {
        pattern_matching_result = 0;
        a = attribute.fields[0];
    }
    switch (pattern_matching_result) {
        case 0: {
            return a;
        }
    }
}

export function getUbo(name, data) {
    const matchValue = tryFind_1((u) => (Data_GlUboData__get_Name(u) === name), data.Ubos);
    if (matchValue == null) {
        throw (new Error(toText(interpolate("Ubo \u0027%P()\u0027 not found", [name]))));
    }
    else {
        const ubo = matchValue;
        return ubo;
    }
}

export function tryGetUbo(name, data) {
    return tryFind_1((u) => (Data_GlUboData__get_Name(u) === name), data.Ubos);
}

export function allUniforms(data) {
    return append(data.Uniforms, collect((ubo) => ubo.Uniforms, data.Ubos));
}

export function allAttributes(data) {
    const getAttributes = (data_1) => delay(() => append(singleton(data_1), delay(() => collect(getAttributes, data_1.ChildAttributes))));
    return collect((arg) => getAttributes(getAttributeData(arg)), data.Attributes);
}

export function getUniform(name, data) {
    const matchValue = tryFind_2((u) => (Data_GlUniformData__get_Name(u) === name), allUniforms(data));
    if (matchValue == null) {
        throw (new Error(toText(interpolate("Uniform \u0027%P()\u0027 not found", [name]))));
    }
    else {
        const uniform = matchValue;
        return uniform;
    }
}

export function tryGetUniform(name, data) {
    return tryFind_2((u) => (Data_GlUniformData__get_Name(u) === name), allUniforms(data));
}

export function getAttribute(name, data) {
    const matchValue = tryFind_2((u) => (Data_GlAttributeData__get_Name(u) === name), allAttributes(data));
    if (matchValue == null) {
        throw (new Error(toText(interpolate("Attribute \u0027%P()\u0027 not found", [name]))));
    }
    else {
        const attribute = matchValue;
        return attribute;
    }
}

export function tryGetAttribute(name, data) {
    return tryFind_2((u) => (Data_GlAttributeData__get_Name(u) === name), allAttributes(data));
}

export function getTexture(name, data) {
    const matchValue = tryFind_1((u) => (u.Name === name), data.Textures);
    if (matchValue == null) {
        throw (new Error(toText(interpolate("Texture \u0027%P()\u0027 not found", [name]))));
    }
    else {
        const texture = matchValue;
        return texture;
    }
}

export function tryGetTexture(name, data) {
    return tryFind_1((u) => (u.Name === name), data.Textures);
}

export function objects(data) {
    return delay(() => collect((l) => l.Objects, reverse(data.Layers)));
}

export function getCamera(name, data) {
    const matchValue = tryFind_1((o) => (o.Name === name), data.Cameras);
    if (matchValue == null) {
        const msg = toText(interpolate("Camera \u0027%P()\u0027 not found", [name]));
        throw (new Error(msg));
    }
    else {
        const camera = matchValue;
        return camera;
    }
}

export function tryGetCamera(name, data) {
    return tryFind_1((u) => (u.Name === name), data.Cameras);
}

export function getObject(name, data) {
    const matchValue = tryFind_2((o) => (o.Name === name), objects(data));
    if (matchValue == null) {
        const msg = toText(interpolate("Object \u0027%P()\u0027 not found", [name]));
        throw (new Error(msg));
    }
    else {
        const object = matchValue;
        return object;
    }
}

export function tryGetObject(name, data) {
    return tryFind_2((u) => (u.Name === name), objects(data));
}

export function dirtyScene(data) {
    data.IsDirty = true;
    data.Canvas.IsDirty = true;
}

export function dirtyObject(data) {
    data.IsDirty = true;
    dirtyScene(data.Scene);
}

export function dirtyModel(data) {
    data.IsModelDirty = true;
    dirtyObject(data);
}

export function dirtyParallax(data) {
    if (data.ParallaxCamera != null) {
        data.IsParallaxDirty = true;
        dirtyObject(data);
    }
}

export function objectRecalcNeeded(data) {
    data.RecalcNeeded = true;
    dirtyObject(data);
}

export function nextUboBufferIndex(data) {
    const loop = (i_mut, ubos_mut) => {
        loop:
        while (true) {
            const i = i_mut, ubos = ubos_mut;
            if (ubos.tail != null) {
                const t = ubos.tail;
                const h = ubos.head;
                if (i < h.Location) {
                    return i | 0;
                }
                else {
                    i_mut = (i + 1);
                    ubos_mut = t;
                    continue loop;
                }
            }
            else {
                return i | 0;
            }
            break;
        }
    };
    return loop(0, sortBy((u) => u.Location, data.Ubos, {
        Compare: comparePrimitives,
    })) | 0;
}

export function nextTextureIndex(data) {
    const loop = (i_mut, ubos_mut) => {
        loop:
        while (true) {
            const i = i_mut, ubos = ubos_mut;
            if (ubos.tail != null) {
                const t = ubos.tail;
                const h = ubos.head;
                if (i < h.Index) {
                    return i | 0;
                }
                else {
                    i_mut = (i + 1);
                    ubos_mut = t;
                    continue loop;
                }
            }
            else {
                return i | 0;
            }
            break;
        }
    };
    return loop(0, sortBy((u) => u.Index, data.Textures, {
        Compare: comparePrimitives,
    })) | 0;
}

export function addScene(creator, data) {
    const sceneData = creator(empty(), data);
    data.Scenes = append_1(data.Scenes, [sceneData]);
    return sceneData;
}

export function removeScene(scene, data) {
    data.Scenes = data.Scenes.filter((o) => (!(o === scene)));
    data.IsDirty = true;
    return data;
}

export function addCamera(creator, data) {
    const camData = creator(empty(), data);
    data.Cameras = append_2(data.Cameras, singleton_1(camData));
    dirtyScene(data);
    return camData;
}

export function removeCamera(cam, data) {
    data.Cameras = filter((o) => (!(o === cam)), data.Cameras);
    dirtyScene(data);
    return data;
}

export function addObject(creator, data) {
    const objData = creator(empty(), data);
    const layer = data.Layers[objData.Layer];
    layer.Objects = append_2(layer.Objects, singleton_1(objData));
    dirtyScene(data);
    return objData;
}

export function removeObject(globj, data) {
    const layer = data.Layers[globj.Layer];
    layer.Objects = filter((o) => (!(o === globj)), layer.Objects);
    dirtyScene(data);
    return data;
}

export function addUbo(creator, data) {
    const uboData = creator(empty(), data);
    data.Ubos = append_2(data.Ubos, singleton_1(uboData));
    dirtyObject(data);
    return uboData;
}

export function addUniform(creator, data) {
    const uniformData = creator(empty(), data);
    data.Uniforms = append_2(data.Uniforms, singleton_1(uniformData));
    dirtyObject(data);
    return uniformData;
}

export function addAttribute(creator, data) {
    const attributeData = creator(empty(), data);
    data.Attributes = append_2(data.Attributes, singleton_1(attributeData));
    dirtyObject(data);
    return attributeData;
}

export function addTexture(creator, data) {
    const textureData = creator(empty(), data);
    data.Textures = append_2(data.Textures, singleton_1(textureData));
    dirtyObject(data);
    return textureData;
}

export function addUboUniform(creator, data) {
    const uniformData = creator(empty(), data, data.ParentObject);
    data.Uniforms = cons(uniformData, data.Uniforms);
    return uniformData;
}

export function objectDebugName(globj) {
    if (!isNullOrWhiteSpace(globj.Name)) {
        return globj.Name;
    }
    else if (globj.Parent == null) {
        return "shared";
    }
    else {
        return "";
    }
}

File: src/webgl_data/glcommon.fs
namespace Wil.Webgl.Data

open System
open Wil.Core.Utils

module GlCommon =
  let splitName (name: string) =
    let pair = name.Split('.')
    (pair.[0], pair.[1])

  let getScene name (data: GlCanvasData) =
    match data.Scenes |> Array.tryFind (fun u -> u.Name = name) with
    | Some scene -> scene
    | None -> raise (exn $"Scene '{name}' not found")

  let allObjects (data: GlCanvasData) =
    let sceneObjects (data: GlSceneData) =
      data.Layers |> Seq.collect (fun l -> l.Objects)
      
    data.Scenes |> Seq.collect sceneObjects

  let getAttributeData (attribute: GlRootAttribute) =
    match attribute with
    | SingleAttribute a
    | InterleaveAttribute a -> a

  let getUbo name (data: GlObjData) =
    match data.Ubos |> List.tryFind (fun u -> u.Name = name) with
    | Some ubo -> ubo
    | None -> raise (exn $"Ubo '{name}' not found")

  let tryGetUbo name (data: GlObjData) =
    data.Ubos |> List.tryFind (fun u -> u.Name = name)

  let allUniforms (data: GlObjData) =
    data.Ubos
    |> Seq.collect (fun ubo -> ubo.Uniforms)
    |> Seq.append data.Uniforms

  let allAttributes (data: GlObjData) =
    let rec getAttributes data =
      seq {
        yield data
        yield! data.ChildAttributes |> Seq.collect getAttributes
      }
 
    data.Attributes |> Seq.collect (getAttributeData >> getAttributes)

  let getUniform name (data: GlObjData) =
    match data |> allUniforms |> Seq.tryFind (fun u -> u.Name = name) with
    | Some uniform -> uniform
    | None -> raise (exn $"Uniform '{name}' not found")

  let tryGetUniform name (data: GlObjData) =
    data |> allUniforms |> Seq.tryFind (fun u -> u.Name = name)

  let getAttribute name (data: GlObjData) =
    match data |> allAttributes |> Seq.tryFind (fun u -> u.Name = name) with
    | Some attribute -> attribute
    | None -> raise (exn $"Attribute '{name}' not found")

  let tryGetAttribute name (data: GlObjData) =
    data |> allAttributes |> Seq.tryFind (fun u -> u.Name = name)

  let getTexture name (data: GlObjData) =
    match data.Textures |> List.tryFind (fun u -> u.Name = name) with
    | Some texture -> texture
    | None -> raise (exn $"Texture '{name}' not found")

  let tryGetTexture name (data: GlObjData) =
    data.Textures |> List.tryFind (fun u -> u.Name = name)

  let objects (data: GlSceneData) =
    seq { yield! data.Layers |> Seq.rev |> Seq.collect (fun l -> l.Objects) }

  let getCamera name (data: GlSceneData) =
    match data.Cameras |> List.tryFind (fun o -> o.Name = name) with
    | Some camera -> camera
    | None -> throw $"Camera '{name}' not found"

  let tryGetCamera name (data: GlSceneData) =
    data.Cameras |> List.tryFind (fun u -> u.Name = name)

  let getObject name (data: GlSceneData) =
    match data |> objects |> Seq.tryFind (fun o -> o.Name = name) with
    | Some object -> object
    | None -> throw $"Object '{name}' not found"

  let tryGetObject name (data: GlSceneData) =
    data |> objects |> Seq.tryFind (fun u -> u.Name = name)

  let dirtyScene (data: GlSceneData) =
    data.IsDirty <- true
    data.Canvas.IsDirty <- true

  let dirtyObject (data: GlObjData) =
    data.IsDirty <- true
    dirtyScene data.Scene

  let dirtyModel (data: GlObjData) =
    data.IsModelDirty <- true
    dirtyObject data

  let dirtyParallax (data: GlObjData) =
    match data.ParallaxCamera with
    | None -> ()
    | Some _ ->
        data.IsParallaxDirty <- true
        dirtyObject data

  let objectRecalcNeeded (data: GlObjData) =
    data.RecalcNeeded <- true
    dirtyObject data

  let nextUboBufferIndex (data: GlObjData) =
    let rec loop i (ubos: GlUboData list) =
      match ubos with
      | [] -> i
      | h::t -> if i < h.Location then i else loop (i + 1) t
    loop 0 (data.Ubos |> List.sortBy (fun u -> u.Location))

  let nextTextureIndex (data: GlObjData) =
    let rec loop i (ubos: GlTextureData list) =
      match ubos with
      | [] -> i
      | h::t -> if i < h.Index then i else loop (i + 1) t
    loop 0 (data.Textures |> List.sortBy (fun u -> u.Index))

  let addScene (creator: GlSceneFactory) (data: GlCanvasData) =
    let sceneData = creator [] data
    data.Scenes <- Array.append data.Scenes [| sceneData |]
    sceneData

  let removeScene (scene: GlSceneData) (data: GlCanvasData) =
    data.Scenes <- data.Scenes |> Array.filter (fun o -> not <| Object.ReferenceEquals(o, scene))
    data.IsDirty <- true
    data

  let addCamera (creator: GlCameraFactory) (data: GlSceneData) =
    let camData = creator [] data
    data.Cameras <- data.Cameras @ [camData]
    dirtyScene data
    camData

  let removeCamera (cam: GlObjData) (data: GlSceneData) =
    data.Cameras <- data.Cameras |> List.filter (fun o -> not <| Object.ReferenceEquals(o, cam))
    dirtyScene data
    data

  let addObject (creator: GlObjectFactory) (data: GlSceneData) =
    let objData = creator [] data
    let layer = data.Layers.[objData.Layer]
    layer.Objects <- layer.Objects @ [objData]
    dirtyScene data
    objData

  let removeObject (globj: GlObjData) (data: GlSceneData) =
    let layer = data.Layers.[globj.Layer]
    layer.Objects <- layer.Objects |> List.filter (fun o -> not <| Object.ReferenceEquals(o, globj))
    dirtyScene data
    data

  let addUbo (creator: GlUboFactory) (data: GlObjData) =
    let uboData = creator [] data
    data.Ubos <- data.Ubos @ [uboData]
    dirtyObject data
    uboData

  let addUniform (creator: GlUniformFactory) (data: GlObjData) =
    let uniformData = creator [] data
    data.Uniforms <- data.Uniforms @ [uniformData]
    dirtyObject data
    uniformData

  let addAttribute (creator: GlRootAttributeFactory) (data: GlObjData) =
    let attributeData = creator [] data
    data.Attributes <- data.Attributes @ [attributeData]
    dirtyObject data
    attributeData

  let addTexture (creator: GlTextureFactory) (data: GlObjData) =
    let textureData = creator [] data
    data.Textures <- data.Textures @ [textureData]
    dirtyObject data
    textureData

  let addUboUniform (creator: GlUboUniformFactory) (data: GlUboData) =
    let uniformData = creator [] data data.ParentObject
    data.Uniforms <- uniformData :: data.Uniforms
    uniformData

  let objectDebugName (globj: GlObjData) =
    if not <| String.IsNullOrWhiteSpace(globj.Name)
    then globj.Name
    elif globj.Parent.IsNone
    then "shared"
    else ""

